### PR TITLE
Workflow: Optional Encounter ID and Search By URL expansion

### PR DESCRIPTION
**Pull request by vitorpamplona**
Two new features in the PR: 
1. Enables the call to generate a care plan without an encounter. 
2. Restructures FhirDal to support search by URL in any ResourceType (Previously, only Measure and Library were allowed). 

The same test case covers both features.

**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #[issue number]

**Description**
Clear and concise code change description. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt [hunk 1]
```diff
@@ -237,7 +237,15 @@ class FhirOperator(fhirContext: FhirContext, fhirEngine: FhirEngine) {
     )
   }
 
-  fun generateCarePlan(planDefinitionId: String, patientId: String, encounterId: String): CarePlan {
+  fun generateCarePlan(planDefinitionId: String, patientId: String): CarePlan {
```
Comment (omarismail94) [inline workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt:240]: thanks @vitorpamplona ! Can you write a quick test to cover this function please?

**Review by omarismail94 — COMMENTED** [review]
Just a small addition!

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt (hunk 1 above)
Comment (vitorpamplona) [reply workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt:240]: While making this test case, I found another bug. I will update this PR soon

**Review by omarismail94 — APPROVED** [review]
Great change @vitorpamplona !

**Comment by omarismail94** [comment]
@vitorpamplona could you merge with master so the PR gets merged?

**Pull request merged**